File: app/src/main/java/com/nvvi9/giphyme/di/DatabaseModule.java
package com.nvvi9.giphyme.di;

import android.app.Application;

import androidx.room.Room;

import com.nvvi9.giphyme.db.GiphyMeDatabase;
import com.nvvi9.giphyme.db.dao.GifInfoDao;
import com.nvvi9.giphyme.db.dao.GifInfoRemoteKeysDao;
import com.nvvi9.giphyme.db.dao.GifPreviewDao;

import javax.inject.Singleton;

import dagger.Module;
import dagger.Provides;
import dagger.hilt.InstallIn;
import dagger.hilt.components.SingletonComponent;

@InstallIn(SingletonComponent.class)
@Module
public class DatabaseModule {

    @Provides
    @Singleton
    GiphyMeDatabase provideDatabase(Application application) {
        return Room.databaseBuilder(application, GiphyMeDatabase.class, "GiphyME.db")
                .build();
    }

    @Provides
    GifPreviewDao provideGifPreviewDao(GiphyMeDatabase database) {
        return database.getGifPreviewDao();
    }

    @Provides
    GifInfoRemoteKeysDao provideGifInfoRemoteKeysDao(GiphyMeDatabase database) {
        return database.getGifInfoRemoteKeysDao();
    }

    @Provides
    GifInfoDao provideGifInfoDao(GiphyMeDatabase database) {
        return database.gitGifInfoDao();
    }
}

File: app/src/main/java/com/nvvi9/giphyme/di/DatabaseModule.kt
package com.nvvi9.giphyme.di

import android.app.Application
import androidx.room.Room.databaseBuilder
import com.nvvi9.giphyme.db.GiphyMeDatabase
import com.nvvi9.giphyme.db.dao.GifInfoDao
import com.nvvi9.giphyme.db.dao.GifInfoRemoteKeysDao
import com.nvvi9.giphyme.db.dao.GifPreviewDao
import dagger.Module
import dagger.Provides
import dagger.hilt.InstallIn
import dagger.hilt.components.SingletonComponent
import javax.inject.Singleton

@InstallIn(SingletonComponent::class)
@Module
class DatabaseModule {
    @Provides
    @Singleton
    fun provideDatabase(application: Application?): GiphyMeDatabase =
        databaseBuilder(application!!, GiphyMeDatabase::class.java, "GiphyME.db")
            .build()


    @Provides
    fun provideGifPreviewDao(database: GiphyMeDatabase): GifPreviewDao =
        database.getGifPreviewDao()

    @Provides
    fun provideGifInfoRemoteKeysDao(database: GiphyMeDatabase): GifInfoRemoteKeysDao =
        database.getGifInfoRemoteKeysDao()

    @Provides
    fun provideGifInfoDao(database: GiphyMeDatabase): GifInfoDao =
        database.gitGifInfoDao()
}